lumicks/pylake/piezo_tracking/piezo_tracking.py

```diff
@@ -0,0 +1,166 @@
+        baseline_force1 : pylake.ForceBaseline
+            Baseline for force1
+        baseline_force2 : pylake.ForceBaseline
+            Baseline for force2
+        signs : tuple(float, float)
+            Sign convention for forces (e.g. (1, -1) indicates that force2 is negative).
+        """
+        if len(signs) != 2:
+            raise ValueError(
+                "Argument `signs` should be a tuple of two floats reflecting the sign for each "
+                "channel."
+            )
+        for sign in signs:
+            if abs(sign) != 1:
+                raise ValueError("Each sign should be either -1 or 1.")
+
+        self.trap_calibration = trap_calibration
+        self.baseline_force1 = baseline_force1
+        self.baseline_force2 = baseline_force2
+        self._signs = signs
+
+    def valid_range(self):
+        """Returns the mirror position range in which the piezo tracking is valid"""
+        calibration_items = (self.trap_calibration, self.baseline_force1, self.baseline_force2)
+        return np.min(np.stack([r.valid_range() for r in calibration_items]), axis=0)
+
+    def piezo_track(self, trap_position, force1, force2, trim=True, downsampling_factor=None):
+        """Obtain piezo distance and baseline corrected forces
+
+        Parameters
+        ----------
+        trap_position : pylake.channel.Slice
+            Trap position.
+        force1 : pylake.channel.Slice
+            First force channel to use for piezo tracking.
+        force2 : pylake.channel.Slice
+            Second force channel to use for piezo tracking.
+        trim : bool
+            Trim regions outside the calibration range.
+        downsampling_factor : Optional[int]
+            Downsampling factor.
+        """
+        if downsampling_factor:
+            trap_position, force1, force2 = (
+                x.downsampled_by(downsampling_factor) for x in (trap_position, force1, force2)
+            )
+
+        trap_trap_dist = self.trap_calibration(trap_position)
+        bead_displacements = 1e-3 * sum(
+            sign * force / force.calibration[0]["kappa (pN/nm)"]
+            for force, sign in zip((force1, force2), self._signs)
+        )
+
+        piezo_distance = trap_trap_dist - bead_displacements
+        corrected_force1 = self.baseline_force1.correct_data(force1, trap_position)
+        corrected_force2 = self.baseline_force2.correct_data(force2, trap_position)
+
+        if trim:
+            valid_range = self.valid_range()
+            valid_mask = np.logical_and(
+                valid_range[0] <= trap_position.data, trap_position.data <= valid_range[1]
+            )
+            piezo_distance, corrected_force1, corrected_force2 = (
+                piezo_distance[valid_mask],
+                corrected_force1[valid_mask],
+                corrected_force2[valid_mask],
+            )
+
+        return piezo_distance, corrected_force1, corrected_force2
```

Comment:
It would be nice to also return the time coordinates. Eg when using the pingpong feature in Bluelake, you get many back and forth motions of the trap in one marker. It is then nice to plot the piezo distance over time and look at rupture events.
In force clamp experiments on motor proteins that change the ratio of ssDNA/dsDNA, one would also like to look at the piezo distance over time, for dwell time analysis.